src/H5Cimage.c

```diff
@@ -1287,6 +1287,10 @@ H5C__decode_cache_image_header(const H5F_t *f, H5C_t *cache_ptr, const uint8_t *
     /* Point to buffer to decode */
     p = *buf;
 
+    /* Ensure buffer has enough data for signature comparison */
+    if ((size_t)(*buf + H5C__MDCI_BLOCK_SIGNATURE_LEN - p) > cache_ptr->image_len)
```

Comment:
The logic is generally fine, but a couple of issues here:

- This should really use the `H5_IS_BUFFER_OVERFLOW` macro from H5private.h to be consistent with other decode routines throughout the library
- This is a yet-to-be-fixed case where the size of `buf` should really be passed in as a parameter to the function rather than relying on it from elsewhere. For example, the size of `buf` here is actually `cache_ptr->image_len + 1`, which doesn't really make a difference for this case, but it's easy to see how relying on `cache_ptr->image_len` here can cause trouble.

Reply:
Sorry I should have been more clear; what I meant was that we should update `H5C__decode_cache_image_header()` with an extra size parameter for the size of `buf` and then use that for overflow calculations rather than `cache_ptr->image_len`. This is similar to other decode routines where the functions have both a buffer and size of the buffer parameter. The calculations should essentially be the same since `cache_ptr->image_len + 1` is what the callers will be passing in for the size parameter, but it should be a bit safer than relying on the size from the `cache_ptr` structure.